### PR TITLE
Improve consistent initialization speed and robustness

### PR DESCRIPTION
**Pull request by MarcBerliner**
# Description

The `IDA` solver takes in initial guesses for `y0` and `ydot0`. Currently, we explicitly calculate `y0` using an algebraic solver before passing it to `IDA`. However, we do not estimate `ydot0` - we just set it to `np.zeros_like(y0)` and rely on `IDA` to determine it. This PR makes a few changes to consistent initialization:

1. Analytically compute `ydot0` using the mass matrix inverse. The `IDA` solver is set up only for **implicit** DAEs, so it cannot properly utilize the mass matrix in initializing the system. As a result, it wastes a lot of time re-computing the consistent initialization for `y0` and `ydot0` in `IDACalcIC`. We can improve this by isolating the differential section of our system of equations and explicitly computing `ydot0` for the rhs,
$$\left. \frac{\mathrm{d}y_d^\*}{\mathrm{d}t} \right|_{t_0} = - ( M\_{d,d} ) ^ {-1} \text{rhs}(y_0^\*, t_0) , $$
where the subscript $d$ refers to the equations containing differential terms and $y_0^\*$ is the consistently initialized $y_0$. **This PR reduces the total solve time of a full 1C discharge from 218 ms to 212 ms and reduces the time integrating the system from 45.0 ms to 40.0 ms.** Additionally, this improves the solver robustness because we reduce reliance on `IDACalcIC`, which occasionally fails.
2. Separate `initial_conditions` and `consistent_initialization`. I found that there was not a clear distinction between these ideas in the solver code. In this update, `initial_conditions` refers to the initial values and guesses provided by the solver or from a previous solution. Initial conditions likely will not satisfy the system of equations at `t0`, while `consistent_initialization` refers to the solution that satisfies the system of equations at `t0`. In some cases, we computed the `initial_conditions` more than once for a single solve, which were removed.
3. Various cleanup of the solver codebase. This section will need additional refactoring to unify the API and remove redundancies, but that is beyond the scope of this PR.

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] Optimization (back-end change that speeds up the code)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Review by valentinsulzer — COMMENTED** [review]
Looks good to me at a superficial level. Much overdue clean up and good to know why we've been getting all these IDACalcIC errors. 
I'll let others (@martinjrobins / @jsbrittain ) who have done more on the IDA KLU solver review in more detail

**Review by martinjrobins — CHANGES_REQUESTED** [review]
Thanks @MarcBerliner, this looks great, thank you for your effort on this. In the future, as per the CONTRIBUTING.md can you please make sure to first open an issue before you start on a piece of work so that it can be discussed with the other developers. It would be great also to get your input on https://github.com/pybamm-team/PyBaMM/issues/3910. Please feel free to add additional items that you think are important for the solvers going forward, or edit existing items. I would like this to be an evolving workplan so we can coordinate everyone's work on the solvers.

File: pybamm/solvers/idaklu_solver.py [hunk 1]
```diff
@@ -994,64 +920,212 @@ def _integrate(self, model, t_eval, inputs_dict=None):
...
+
+        y0 = model.y0
+        if isinstance(y0, casadi.DM):
+            y0 = y0.full()
+        y0 = y0.flatten()
+
+        # calculate the time derivatives of the differential equations
+        # for semi-explicit DAEs
+        calc_ydot0 = model.mass_matrix_inv is not None
+        if model.len_rhs > 0 and calc_ydot0:
```
[the remark this answers is not in the document]
Comment (MarcBerliner) [reply pybamm/solvers/idaklu_solver.py]: Good point - we will always have a mass matrix and its inverse if `len_rhs` is nonzero. Removing `calc_ydot0`

**Comment by MarcBerliner** [comment]
Thanks for the quick feedback @martinjrobins. Going forward, I'll make sure to keep you all informed of my planned work. I'll add a few comments to the solver roadmap.

**Review by martinjrobins — APPROVED** [review]
looks good, thanks @MarcBerliner

**Pull request merged**